### PR TITLE
refactor: remove app boilerplate and improvements to cx protocol

### LINKED ISSUE #216: Remove the need for App boilerplate
All CDK apps today have this boilerplate.

JavaScript:

```js
const app = new App(process.argv);
// add stack here
process.stdout.write(app.run());
```

Java:

```java
public static void main(final String argv[]) {
    App app = new App(Arrays.asList(argv));
    // add stack here
    System.out.writeln(app.run());
}
```

The reason we need this boilerplate is due to how the toolkit and CDK apps interact, and the fact that the toolkit is language-agnostic. As far as the toolkit is concerned, CDK apps are executables that adhere to the "cx-api" interface, which is using ARGV to accept inputs and STDOUT as outputs. This is the lowest common denominator across languages (any language out there allows you to write programs that accept ARGV and emit STDOUT).

This is all good an fine, but there's no value in exposing this to users if we can provide a better experience (understanding this protocol is not something that most users will benefit from).

Let's modify the protocol such that instead of ARGV/STDOUT it will pass in the toolkit request via environment variables and the output will be emitted to a temporary working directory. This will make the app boilerplate code look like this:

```ts
const app = new cdk.App();

new MyStack(app, 'stack1');
new YourStack(app, 'stack2');

app.run();
```

Much cleaner.

### PR DESCRIPTION
**Pull request by eladb**
Change the cx protocol to use environment variables and file output
instead of argv/stdout. This simplifies the App boilerplate code.
Now, apps look like this:

    const app = new App();
    // add stacks
    app.run();

The protocol was also simplified to basically always synthesize all
stacks in the app, eliminating the need for a "cx request" altogether.

The toolkit was modified to cache the response of the first successful
execution of the app (after environment contextual is resolved). This
reduces the number of app executions to a maximum of 2 (if env context
is missing) and normally 1. Previoiusly, "list" and "synth" were two
separate commands, so 2 was the normal case.

The protocol now uses the following environment variables:

 * `CDK_OUTDIR` represents the synthesis output directory. The toolkit
   allocates a temporary working directory when it executes the app and
   sets this variable.
 * `CDK_CONTEXT_JSON` is a JSON stringified context object.

At the moment, if `CDK_OUTDIR` is not defined, an error message is printed
indicating the minimum version of the toolkit required to interact with
the app.

Fixes #216

BREAKING CHANGE

This is a major breaking change:

- The `cdk.App` initializer doesn't accept any arguments.
- The `cdk.App#run` method does not return a `string` anymore.

All AWS CDK apps in all languages would need to be modified
to adhere to the new API of the `cdk.App` construct.

Instead of:

    const app = new App(process.argv); // ERROR
    // add stacks
    process.stdout.write(app.run());   // ERROR

The new usage is:

    const app = new App();
    // add stacks
    app.run();

In order to interact with applications written using this
version, the CDK Toolkit must also be update using:

    $ npm i -g aws-cdk

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license.

### CODE REVIEW

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts [hunk 1]
```diff
@@ -1,11 +1,28 @@
 import { cloudformation as applicationautoscaling } from '@aws-cdk/aws-applicationautoscaling';
 import { Role } from '@aws-cdk/aws-iam';
+import iam = require('@aws-cdk/aws-iam');
 import { Construct, PolicyStatement, PolicyStatementEffect, ServicePrincipal } from '@aws-cdk/cdk';
 import { cloudformation as dynamodb } from './dynamodb.generated';
 
 const HASH_KEY_TYPE = 'HASH';
 const RANGE_KEY_TYPE = 'RANGE';
 
+const READ_DATA_ACTIONS = [
```
Comment (RomainMuller) [inline packages/@aws-cdk/aws-dynamodb/lib/table.ts]: Doesn't this belong to a different PR?

File: packages/@aws-cdk/cdk/lib/app.ts [hunk 2]
```diff
@@ -184,61 +147,13 @@ export class App extends Root {
...
-  ${progname}REQUEST
-
-REQUEST is a JSON-encoded request object.
-`;
-  }
-
   private loadContext() {
-    const context = (this.request && this.request.context) || {};
+    const contextJson = process.env[cxapi.CONTEXT_ENV];
+    const context = !contextJson ? { } : JSON.parse(contextJson);
```
Comment (RomainMuller) [inline packages/@aws-cdk/cdk/lib/app.ts]: About as awkward, but doesn't use a ternary: `JSON.parse(contextJson || '{}')`?

File: packages/@aws-cdk/cdk/test/test.app.ts [hunk 3]
```diff
@@ -1,120 +1,94 @@
...
 
-function main(...argv: string[]): any {
-  const app = new App([ 'myprog', ...argv ]);
-
-  // stacks are children of the program
-
-  const stack1 = new Stack(app, 'stack1', { env: { account: '12345', region: 'us-east-1' } });
-  new Resource(stack1, 's1c1', { type: 'DummyResource', properties: { Prop1: 'Prop1' } });
-  const r2 = new Resource(stack1, 's1c2', { type: 'DummyResource', properties: { Foo: 123 } });
+  if (context) {
```
Comment (RomainMuller) [inline packages/@aws-cdk/cdk/test/test.app.ts]: `process.env[cxapi.CONTEXT_ENV] = JSON.stringify(context || {})`?

File: packages/aws-cdk/lib/contextplugins.ts [hunk 4]
```diff
@@ -64,5 +64,6 @@ export async function provideContextValues(missingValues: { [key: string]: Missi
 
     const value = await provider.getValue(query.scope, query.args);
     projectConfig.set(['context', key], value);
+    debug(`Setting "${key}" context to ${value}`);
```
Comment (RomainMuller) [inline packages/aws-cdk/lib/contextplugins.ts]: Might want to `JSON.stringify(value)`, in case it's one of those things that `.toString()`'s to `[Object object]`.

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts (hunk 1 above)
Comment (eladb) [reply packages/@aws-cdk/aws-dynamodb/lib/table.ts]: ha! i wondered where these changes disappeared. Yes, of course. This needs to go to a separate PR.

File: packages/@aws-cdk/cdk/lib/app.ts (hunk 2 above)
Comment (eladb) [reply packages/@aws-cdk/cdk/lib/app.ts]: Prepare the ternary

File: packages/aws-cdk/lib/contextplugins.ts (hunk 4 above)
Comment (eladb) [reply packages/aws-cdk/lib/contextplugins.ts]: 👍

File: packages/@aws-cdk/cdk/test/test.app.ts (hunk 3 above)
Comment (eladb) [reply packages/@aws-cdk/cdk/test/test.app.ts]: I actually want to remove this environment variable altogether. I tried to assigned `undefined` but for some weird reason, it will assign an `"undefined"` string to it instead of remove it.
